### ISSUE TITLE
[Gap Analysis] Implement Smart Distributed Training (FSDP-Inspired)

### ISSUE THREAD

**Issue by ooples** [ooples/AiDotNet#309]
### User Story

> As a researcher, I want to train models that are too large to fit on a single GPU, so that I can scale my experiments to state-of-the-art architectures and large datasets without being limited by hardware memory constraints.

---

### **Problem & Proposed Solution**

Training large models is impossible on a single GPU. Frameworks like PyTorch's Fully Sharded Data Parallelism (FSDP) solve this by sharding model parameters, gradients, and optimizer states across multiple GPUs. However, FSDP can be complex to configure and may suffer from CPU overhead due to numerous small communication operations.

This plan outlines a **smarter, simpler FSDP-inspired framework** for AiDotNet. It will improve upon known weaknesses by introducing **automated parameter grouping** to reduce communication overhead and providing a **radically simple API**.

---

### **Phase 1: Core Communication Abstraction**

**Goal:** Create a backend-agnostic wrapper for fundamental distributed communication primitives.

#### **AC 1.1: Research and Implement Communication Backend (3 points)**
**Requirement:** Select and integrate a high-performance communication library (e.g., MPI).

-   [ ] Research available .NET libraries for collective communication (e.g., `Microsoft.MPI`, `MPICH.Net`). Select the one with the best performance and cross-platform support.
-   [ ] Create a new project `src/AiDotNet.Distributed` for all distributed training components.

#### **AC 1.2: Create the `CommunicationManager` (5 points)**
**Requirement:** A static class that abstracts away the chosen backend, providing a clean internal API.

-   [ ] Create a file: `src/Distributed/CommunicationManager.cs`.
-   [ ] Implement static methods that wrap the chosen backend:
    -   `[ ] public static void Initialize(string[] args)`
    -   `[ ] public static int GetRank()`
    -   `[ ] public static int GetWorldSize()`
    -   `[ ] public static void Barrier()`
    -   `[ ] public static Tensor<T> AllReduce(Tensor<T> tensor, ReduceOp op = ReduceOp.Sum)`
    -   `[ ] public static Tensor<T>[] AllGather(Tensor<T> tensor)`

---

### **Phase 2: Fully Sharded Model and Optimizer**

**Goal:** Implement the core logic for sharding model parameters and optimizer states.

#### **AC 2.1: Create `ShardedModel<T>` Wrapper (13 points)**
**Requirement:** A class that wraps an existing `IModel<T>` and shards its parameters across devices.

-   [ ] Create a file: `src/Distributed/ShardedModel.cs`.
-   [ ] The `ShardedModel<T>` constructor will take an `IModel<T>` and a `ShardingStrategy`.
-   [ ] **Parameter Sharding:** On initialization, each process (rank) will iterate through the model's parameters. It will drop all parameters that do not belong to its shard, keeping only a fraction of the total model to save memory. The original parameters are preserved on a meta-device (CPU) for model saving.
-   [ ] **Forward Pass:** The `Forward` method will implement the FSDP logic:
    1.  [ ] Before executing a layer, it will perform an `AllGather` operation to retrieve the necessary parameters for that layer from all other ranks.
    2.  [ ] Execute the layer's forward pass with the full, reconstituted parameters.
    3.  [ ] Immediately discard the gathered parameters to free memory.
-   [ ] **Backward Pass:** The `Backward` method will use a custom gradient hook. After local gradients are computed, it will perform an `AllReduce` operation to sum the gradients across all ranks. The result is that each rank has the correct, globally-summed gradient for the parameters it is responsible for.

#### **AC 2.2: Create `ShardedOptimizer<T>` Wrapper (8 points)**
**Requirement:** An optimizer that wraps a standard AiDotNet optimizer and is aware of the sharded parameters.

-   [ ] Create a file: `src/Distributed/ShardedOptimizer.cs`.
-   [ ] The `ShardedOptimizer<T>` will wrap an existing `IOptimizer<T>`.
-   [ ] The `Step()` method will only update the local shard of parameters that each rank is responsible for. Since gradients have already been reduced by the `ShardedModel`, the optimizer logic remains simple.

---

### **Phase 3: "Smart Sharding" and Simplified API (The Improvement)**

**Goal:** Address FSDP weaknesses with intelligent automation and a simple user-facing API.

#### **AC 3.1: Implement Automated Parameter Grouping (8 points)**
**Requirement:** Reduce communication overhead by automatically grouping small parameters into larger blocks before sharding.

-   [ ] In the `ShardedModel` constructor, before sharding, implement a `ParameterAnalyzer`.
-   [ ] This analyzer will iterate through the model's parameters and identify small parameters (e.g., biases, small embedding layers) based on a configurable size threshold.
-   [ ] It will automatically group these small parameters into larger, contiguous blocks.
-   [ ] The sharding logic will then treat each block as a single unit. This significantly reduces the number of `AllGather` calls, which is a major performance bottleneck in naive FSDP.

#### **AC 3.2: Create the `.AsDistributed()` API (5 points)**
**Requirement:** A single, simple extension method to make a model distributed.

-   [ ] Create an extension method `public static IModel<T> AsDistributed<T>(this IModel<T> model)`.
-   [ ] This method will handle all the setup:
    1.  [ ] Initialize the `CommunicationManager`.
    2.  [ ] Wrap the model in the `ShardedModel<T>` wrapper.
    3.  [ ] Return the wrapped, distributed model.
-   [ ] The user's code should be as simple as: `var model = new MyModel().AsDistributed();`

---

### **Phase 4: Integration and Testing**

**Goal:** Verify the correctness and performance of the distributed framework.

#### **AC 4.1: Create Launcher Script (3 points)**
**Requirement:** A script to launch a multi-process training job.

-   [ ] Create a PowerShell or bash script (`run_distributed.ps1`) that uses `mpiexec` (or the chosen backend's runner) to launch the same C# training application on multiple processes.

#### **AC 4.2: End-to-End Integration Test (8 points)**
**Requirement:** A test case that proves a model trained in a distributed fashion is numerically identical to one trained on a single device.

-   [ ] Create a training application that can be run by the launcher script.
-   [ ] The application will train a simple model on a small, fixed dataset.
-   [ ] **Test Logic:**
    1.  [ ] Run the training script on a single process and save the final model weights.
    2.  [ ] Run the training script on two processes using the `.AsDistributed()` API and save one of the final models.
    3.  [ ] Write a separate assertion program that loads both saved models and asserts that their parameters are numerically identical (within a small tolerance).

---

### **Definition of Done**

-   [ ] All checklist items are complete.
-   [ ] A developer can take a standard AiDotNet model and, by adding `.AsDistributed()`, train it on multiple GPUs/machines.
-   [ ] The framework intelligently shards the model, gradients, and optimizer states to save memory.
-   [ ] The "Smart Sharding" feature is implemented to improve performance over naive FSDP.
-   [ ] End-to-end tests prove numerical correctness.
---

## ⚠️ CRITICAL ARCHITECTURAL REQUIREMENTS

**Before implementing this user story, you MUST review:**
- **📋 Full Requirements:** [`.github/USER_STORY_ARCHITECTURAL_REQUIREMENTS.md`](../.github/USER_STORY_ARCHITECTURAL_REQUIREMENTS.md)
- **📐 Project Rules:** [`.github/PROJECT_RULES.md`](../.github/PROJECT_RULES.md)

#### 1. INumericOperations<T> Usage (CRITICAL)
- [ ] Include `protected static readonly INumericOperations<T> NumOps = MathHelper.GetNumericOperations<T>();` in base class
- [ ] NEVER hardcode `double`, `float`, or specific numeric types - use generic `T`
- [ ] NEVER use `default(T)` - use `NumOps.Zero` instead
- [ ] Use `NumOps.Zero`, `NumOps.One`, `NumOps.FromDouble()` for values
- [ ] Use `NumOps.Add()`, `NumOps.Multiply()`, etc. for arithmetic
- [ ] Use `NumOps.LessThan()`, `NumOps.GreaterThan()`, etc. for comparisons

#### 2. Inheritance Pattern (REQUIRED)
- [ ] Create `I{FeatureName}.cs` in `src/Interfaces/` (root level, NOT subfolders)
- [ ] Create `{FeatureName}Base.cs` in `src/{FeatureArea}/` inheriting from interface
- [ ] Create concrete classes inheriting from Base class (NOT directly from interface)

#### 3. PredictionModelBuilder Integration (REQUIRED)
- [ ] Add private field: `private I{FeatureName}<T>? _{featureName};` to `PredictionModelBuilder.cs`
- [ ] Add Configure method taking ONLY interface (no parameters):
  ```csharp
  public IPredictionModelBuilder<T, TInput, TOutput> Configure{FeatureName}(I{FeatureName}<T> {featureName})
  {
      _{featureName} = {featureName};
      return this;
  }
  ```
- [ ] Use feature in `Build()` with default: `var {featureName} = _{featureName} ?? new Default{FeatureName}<T>();`
- [ ] Verify feature is ACTUALLY USED in execution flow

#### 4. Beginner-Friendly Defaults (REQUIRED)
- [ ] Constructor parameters with defaults from research/industry standards
- [ ] Document WHY each default was chosen (cite papers/standards)
- [ ] Validate parameters and throw `ArgumentException` for invalid values

#### 5. Property Initialization (CRITICAL)
- [ ] NEVER use `default!` operator
- [ ] String properties: `= string.Empty;`
- [ ] Collections: `= new List<T>();` or `= new Vector<T>(0);`
- [ ] Numeric properties: appropriate default or `NumOps.Zero`

#### 6. Class Organization (REQUIRED)
- [ ] One class/enum/interface per file
- [ ] ALL interfaces in `src/Interfaces/` (root level)
- [ ] Namespace mirrors folder structure (e.g., `src/Regularization/` → `namespace AiDotNet.Regularization`)

#### 7. Documentation (REQUIRED)
- [ ] XML documentation for all public members
- [ ] `<b>For Beginners:</b>` sections with analogies and examples
- [ ] Document all `<param>`, `<returns>`, `<exception>` tags
- [ ] Explain default value choices

#### 8. Testing (REQUIRED)
- [ ] Minimum 80% code coverage
- [ ] Test with multiple numeric types (double, float)
- [ ] Test default values are applied correctly
- [ ] Test edge cases and exceptions
- [ ] Integration tests for PredictionModelBuilder usage

---

**⚠️ Failure to follow these requirements will result in repeated implementation mistakes and PR rejections.**

**See full details:** [`.github/USER_STORY_ARCHITECTURAL_REQUIREMENTS.md`](../.github/USER_STORY_ARCHITECTURAL_REQUIREMENTS.md)
